### PR TITLE
Feature/#43 add performance test example

### PR DESCRIPTION
**Pull request by Rabid-Dan**
This addresses issue #43 

I've ported over the performance test setup we had in OFLC and built a new variant that'll get included by default in new apps.
One caveat to the variant is that it requires lighthouse to be set up which isn't included in this PR because the accessibility test variant sets this up first. 
It would be nice for the performance variant to be self contained and not reliant on the accessibility one but I'm not super well versed in templating so if theres a way I can alter them to both have the lighthouse setup without accidentally adding it twice then please do let me know.

### CODE REVIEW

File: variants/performance/spec/support/shared_examples/a_performant_page.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+shared_examples "a performant page", uses_javascript: true do
+  it "with a minimum passing score of 95" do
+    expect(page).to pass_lighthouse_audit(:performance, score: 95)
```
Comment (mischa-s) [inline variants/performance/spec/support/shared_examples/a_performant_page.rb:3]: :racing_car:

File: variants/performance/spec/performance_helper.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  # performance specs against.  Without this webpacker doesnt zip the js packs
+  # and the performance specs show vastly different results to a live
+  # environment.
+  config.before(:suite) do
+    puts "*" * 80
+    puts "NOTICE: This Rspec run includes performance tests from spec/performance/"
+    puts "NOTICE: We are pre-compiling assets to get realistic results from the performance tests"
+    puts "NOTICE: Pre-compilation can take a while ..."
+
+    # Use FileUtils rather than shelling out so that this works on Windows & Unixen
```
Comment (fractalcactus) [inline variants/performance/spec/performance_helper.rb:27]: I think the intended comment should be `Windows & Unix environments` not `Windows & Unixen`?

Comment (Rabid-Dan) [reply variants/performance/spec/performance_helper.rb:27]: oh thanks good catch!

**Comment by Rabid-Dan** [comment]
the failing test on this will be fixed when either #113 or #117 land and I dont really want to fix it a third time so I'll just wait to merge this until either of those are through

**Comment by joshmcarthur** [comment]
Unfortunately there isn't a lot of context in #43, but I think we need a bit more context from @eoinkelly (the issue reporter) on this one to clarify the definition of 'performance test'. My recollection of the performance test that we've done in the past have related to more of a benchmark/server-side response time, rather than frontend performance.

**Comment by Rabid-Dan** [comment]
@joshmcarthur thats a really good point, my experience with performance testing is pretty limited to this implementation so I didn't even think about it being a backend benchmarking test.

**Comment by joshmcarthur** [comment]
As I said, I'm unsure either. I'd be more tempted to take what you've done and bundle it all in to a lighthouse variant that has sub-questions for whether to add accessibility, performance, PWA, to avoid to dependency issue you've identified.

**Comment by Rabid-Dan** [comment]
That sounds like a pretty good implementation. I think if this is the type of performance test Eoin was thinking of then I'll re-work this PR to setup the first bit of lighthouse variant setup with just the accessibility and performance tests baked into it. I feel like we probably don't need questions for it at this point though. Accessibility testing should definitely be the norm and I think performance testing should probably be treated pretty similarly if we're going to build a culture of both of them being important considerations in every project.

**Comment by joshmcarthur** [comment]
@danmorrisonNZ I agree, but I also wonder if some of this effort should be directed into github.com/ackama/lighthouse-matchers. At the moment, I believe lighthouse-matchers runs those checks pretty inefficiently, so I don't want to rush to add checks "just cos"

**Comment by Rabid-Dan** [comment]
@joshmcarthur definitely agree that we should put some effort into the performance of lighthouse-matchers. I've seen it slow test suites down quite noticeably so putting some extra considered time into it in the same way that we're doing with the template here feels like a really worthwhile thing to do. If you've got any thoughts on how we could make it more efficient then I'd be keen to have a crack at contributing to it.

**Comment by eoinkelly** [comment]
This was the kind of perf test I was thinking of :sparkles:. I did a bunch of perf testing using [Jmeter](https://jmeter.apache.org/) on a previous project but that'd be a totally different kind of perf testing.

**Comment by Rabid-Dan** [comment]
@joshmcarthur I quite liked your idea of a lighthouse variant with sub variants for each so I've done that but haven't made them optional yet. Theres not heeapps of issues up for working on for this template at the moment so I'll have a bit of a hack at seeing if I can make the lighthouse matchers stuff more performant today.

**Comment by joshmcarthur** [comment]
Hey @Rabid-Dan. Crap, I think I maybe have bad news for ya. I was talking this over with Eoin this morning and I think we ot to the fact that putting lighthouse into it's own variant probably isn't the way to go since it's putting the tool front-and-center, instead of the feature or problem. 

I think we established that we'd probably have three variants using Lighthouse:

* Performance
* Accessibility
* PWA

Each of these would use Lighthouse, but most would also add something on top - like accessibility also adds axe-matchers, and PWA would add a bunch of offline/web manifest optimisations.

Long story short, probably best to leave this as it is in the performance variant, and if we run into problems with variants clobbering the others' files we can deal with that seperately.

**Comment by Rabid-Dan** [comment]
@joshmcarthur haha oh nooooo. All good, reorganising the files wasnt a big job so I'll just revert the commit and merge it as it was before.
